### PR TITLE
Bug #152: Tests and fixes for numerous breakpoint issues

### PR DESCRIPTION
**Pull request by jonahgraham**
See #152

### CODE REVIEW

**Comment by jonahgraham** [comment]
This is not ready to merge yet. It is here for visibility to other contributors.

File: src/integration-tests/functionBreakpoints.spec.ts [hunk 1]
```diff
@@ -45,7 +45,7 @@ describe('function breakpoints', async () => {
             ],
         });
         await dc.configurationDoneRequest();
-        dc.assertStoppedLocation('function breakpoint', { line: 6 });
+        await dc.assertStoppedLocation('function breakpoint', { line: 6 });
```
Comment (jonahgraham) [inline src/integration-tests/functionBreakpoints.spec.ts]: when I simplified this code the other day I left out the await - cc @thegecko FYI

File: src/GDBDebugSession.ts [hunk 2]
```diff
@@ -333,14 +333,20 @@ export class GDBDebugSession extends LoggingDebugSession {
...
+                        location: `${file}:${vsbp.line}`,
+                        condition: vsbp.condition,
+                        temporary,
+                        ignoreCount,
+                    });
+                    actual.push(createState(vsbp, gdbbp.bkpt));
+                } catch (err) {
+                    actual.push({
+                        verified: false,
+                        message: err.message,
```
Comment (jonahgraham) [inline src/GDBDebugSession.ts]: I don't know if the err.message is sufficient yet - maybe more info is needed?

**Comment by jonahgraham** [comment]
rerun tests

**Comment by jonahgraham** [comment]
:-( There is an unstable test in this PR - passes reliably on my machine (locally and yarn docker:test) and fails most of the time on build machine.

**Comment by jonahgraham** [comment]
> unstable test in this PR 

I think this is the same problem as #145 - just for continueRequest instead of next/step/etc.

**Comment by jonahgraham** [comment]
OK. that worked: CI rerun tests to see if it was a fluke.

**Comment by jonahgraham** [comment]
@thegecko If you want to review this let me know. I'll merge tomorrow if not.

The first three commits are all the new tests, and the other commits fix specific issues. I don't plan to squash these commits, but I'll merge them with a merge commit to keep the related items together.

File: src/GDBDebugSession.ts [hunk 3]
```diff
@@ -407,12 +427,26 @@ export class GDBDebugSession extends LoggingDebugSession {
...
-                this.functionBreakpoints.push(gdbbp.bkpt.number);
-                actual.push(createActual(gdbbp.bkpt));
+            for (const bp of resolved) {
+                if (bp.gdbbp) {
+                    actual.push(createActual(bp.gdbbp));
+                    continue;
+                }
+
+                try {
+                    const vsbp = bp.vsbp;
```
Comment (thegecko) [inline src/GDBDebugSession.ts]: nit: If this is only used once, it doesn't need putting in a separate variable IMO

File: src/GDBDebugSession.ts [hunk 4]
```diff
@@ -280,8 +285,8 @@ export class GDBDebugSession extends LoggingDebugSession {
                     const vsbpCond = vsbp.condition || undefined;
                     const gdbbpCond = gdbbp.cond || undefined;
 
-                    // TODO probably need more thorough checks than just line number
-                    return !!(gdbbp.line && parseInt(gdbbp.line, 10) === vsbp.line
+                    // Check with original-location so that relocated breakpoints are properly matched
+                    return !!(gdbbp['original-location'] === `${file}:${vsbp.line}`
```
Comment (thegecko) [inline src/GDBDebugSession.ts]: I wonder if `original-location` resolves to the same thing across OS's?

File: src/GDBDebugSession.ts [hunk 5]
```diff
@@ -302,25 +307,34 @@ export class GDBDebugSession extends LoggingDebugSession {
                     this.logPointMessages[gdbbp.number] = vsbp.logMessage;
                 }
 
+                let line = 0;
```
Comment (thegecko) [inline src/GDBDebugSession.ts]: 👍

File: src/GDBDebugSession.ts [hunk 6]
```diff
@@ -302,25 +307,34 @@ export class GDBDebugSession extends LoggingDebugSession {
...
-            const actual = existing.map((bp) => createState(bp.vsbp, bp.gdbbp));
+            const actual: DebugProtocol.Breakpoint[] = [];
 
-            existing.forEach((bp) => {
-                if (bp.vsbp.logMessage) {
-                    this.logPointMessages[bp.gdbbp.number] = bp.vsbp.logMessage;
+            for (const bp of resolved) {
+                if (bp.gdbbp) {
+                    actual.push(createState(bp.vsbp, bp.gdbbp));
+                    if (bp.vsbp.logMessage) {
```
Comment (thegecko) [inline src/GDBDebugSession.ts]: We no longer need to add the logmessage here as it is done in `createState()` above.

File: src/GDBDebugSession.ts (hunk 4 above)
Comment (jonahgraham) [reply src/GDBDebugSession.ts]: AFAICT it does - however this code is still wrong as we need to apply standardEscape (GDBBackend) to it so it compares the same. On Windows it is a complicated mess of single, double and quadruple backslashes (e.g. `original-location="d:\\\\cdt\\\\git\\\\cdt-gdb-adapter\\\\src\\\\integration-tests\\\\test-programs\\\\count.c:4"` in the log)

**Comment by jonahgraham** [comment]
I have changed to using `original-location` for matching breakpoints. However the way I implemented it requires GDB >= 7.11. See #159

**Comment by jonahgraham** [comment]
OK - I think I have resolved everything.

I have run this with each version of GDB from 7.11 to 8.3 (using `yarn test:integration --gdb-path gdb.x.y` -- the docker container has all those, and [more](https://github.com/eclipse-cdt/cdt-infra/blob/6ee4556e97973003a9b0a06ced79a7baba023151/docker/cdt-infra-all-gdbs/ubuntu-18.04/Dockerfile#L35), on the PATH)

I am going to merge this (once CI confirms) - if #159 comes back needing < 7.11 I will revisit again. I think this changes I made in fa04aad can be made to work with < 7.11 - but I won't bother unless there is a compelling reason.

**Comment by thegecko** [comment]
Awesome, can we have a release on npm?

**Pull request merged**